File: 4. String Algorithms/Programming-Assignment-1/trie/trie.py
#Uses python3
import sys

# Return the trie built from patterns
# in the form of a dictionary of dictionaries,
# e.g. {0:{'A':1,'T':2},1:{'C':3}}
# where the key of the external dictionary is
# the node ID (integer), and the internal dictionary
# contains all the trie edges outgoing from the corresponding
# node, and the keys are the letters on those edges, and the
# values are the node IDs to which these edges lead.
def build_trie(patterns):
    tree = dict()
    # write your code here

    tree[0] = {}
    counter = 0

    for pattern in patterns:
    	current_node = 0
    	for i in range(0, len(pattern)):
    		current_symbol = pattern[i]
    		if current_symbol in tree[current_node]:
    			current_node = tree[current_node][current_symbol]
    		else:
    			counter += 1
    			tree[current_node][current_symbol] = counter
    			tree[counter] = {}
    			current_node = counter

    return tree


if __name__ == '__main__':
    patterns = sys.stdin.read().split()[1:]
    tree = build_trie(patterns)
    for node in tree:
        for c in tree[node]:
            print("{}->{}:{}".format(node, tree[node][c], c))

File: 4. String Algorithms/Programming-Assignment-1/trie/trie.cpp
#include <string>
#include <iostream>
#include <vector>
#include <map>

using namespace std;

typedef map<char, int> edges;
typedef vector<edges> trie;

trie build_trie(vector<string> & patterns) {
    trie t;
    int counter = 0;
    edges e;
    t.push_back(e);

    for (string pattern : patterns) {
        int current_node = 0;
        for (int i = 0; i < pattern.length(); i++) {
            char c = pattern[i];
            edges edge = t[current_node];
            if (edge.find(c) != edge.end()) {
                current_node = edge[c];
            } else {
                counter++;
                edge[c] = counter;
                current_node = counter;
            }
        }
        t.push_back(edge);
    }

    return t;
}

int main() {
    size_t n;
    std::cin >> n;
    vector<string> patterns;
    for (size_t i = 0; i < n; i++) {
        string s;
        std::cin >> s;
        patterns.push_back(s);
    }

    trie t = build_trie(patterns);
    for (size_t i = 0; i < t.size(); ++i) {
        for (const auto & j : t[i]) {
            std::cout << i << "->" << j.second << ":" << j.first << "\n";
        }
    }

    return 0;
}